Feature: Use python to generate files

WhoBrokeTheBuild opened the pull request:
This work was originally part of the CMake branch, which I am breaking up into smaller PRs
This is being tagged as "Feature:" instead of "Build:" due to how many files it touches, I want to update the minor version number.

Add `opcodes.csv` to be parsed by `gen-*` scripts
Convert `opcbuiltins.h` and `tdishr.h` into generated files, generated by `opcodes.csv`
Update, Rename and Add `gen-*` scripts
Move yacc/lex files out of `yylex/` directories
Update `bootstrap` to call new `gen-*` scripts

WhoBrokeTheBuild committed:
Feature: Use python to generate files

This work was originally part of the CMake branch, which I am breaking up into smaller PRs
This is being tagged as "Feature:" instead of "Build:" due to how many files it touches, I want to update the minor version number.

Add opcodes.csv to be parsed by gen-* scripts
Convert opcbuiltins.h and tdishr.h into generated files, generated by opcodes.csv
Update, Rename and Add gen-* scripts
Move yacc/lex files out of yylex/ directories
Update bootstrap to call new gen-* scripts

WhoBrokeTheBuild committed:
Fixes to support Python 2

zack-vii commented on deploy/gen-include-tdishr.py:
```diff
@@ -0,0 +1,72 @@
+for line in reader:
+    opcode_names.append(line['name'])
+
+output_file = open(OUTPUT_FILENAME, 'wt')
```
you should use a context manager
```python
with open(...) as f:
    f.write(...)
    ...
```
instead of
```python
f = open(...)
f.write(...)
...
f.close()
```

zack-vii commented on deploy/gen-include-tdishr.py:
```diff
@@ -0,0 +1,72 @@
+import csv
+import shutil
+
+from subprocess import *
```
is this import even used? besides you should avoid the usage of `from x import *` its not nice to IDEs and humans ;)

I suggest to autopep8 all gen-*.py files .

WhoBrokeTheBuild replied:
Will do! I actually switched it to `import subprocess` in the other ones, I'll update this one to match.

zack-vii commented on deploy/gen-messages-exceptions.py:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env python
```
this shebang is missing from other files.. and should probably rather be python3.. well you are calling it explicitly with ${PYTHON} so it does not really matter for the auto generation. But for testing and developing it may be nice to be able to quickly call a single generator. In the end it depends on if you want these to be chmod-executable or not. they should be all alike though.

WhoBrokeTheBuild replied:
I generally avoid putting the shebang line, I think this one is from the original file I modified this from. Ideally a human shouldn't be calling this script unless they're updating the script, it should be called by `bootstrap` or from CMake directly after the merge. I'll remove it from this so it's clear you shouldn't be `./` calling these files.

WhoBrokeTheBuild commented:
Thanks so much for the review @zack-vii ! I know how busy you are and I really appreciate it

WhoBrokeTheBuild committed:
Run autopep8 on all gen-* scripts

Rename `gen-python-compound.py` to `gen-python-MDSplus-compound.py` to match the format of:
`gen-include-tdishr.py` -> `include/tdishr.h`
`gen-tdishr-TdiHash.py` -> `tdishr/TdiHash.h`
Which works except for the yacc/lex and messages/exceptions generation scripts.
Remove shebang line from `gen-messages-exceptions.py` and `gen-python-MDSplus-compound.py`
Remove executable permission on those files as well, they should only be called as `python3 deploy/gen-[name].py`

WhoBrokeTheBuild committed:
Refactor, Delete tdishr.h

Actually delete tdishr.h, which should have been done already, it will now be generated by `gen-include-tdishr.py`
Refactor file opens to use `with` statements
Refactor multi-line strings to chains of single-line strings, to better handle indentation

WhoBrokeTheBuild committed:
Fix TdiHash.c generation, Add include/tdishr.h to gitignore

Fix whitespace in intermediate `TdiHash.c.in` file

WhoBrokeTheBuild committed:
Merge branch 'alpha' into slw-cmake-gen-scripts

WhoBrokeTheBuild committed:
Merge branch 'alpha' into slw-cmake-gen-scripts

mwinkel-dev commented on deploy/gen-include-opcbuiltins.py:
```diff
@@ -0,0 +1,41 @@
+        output_file.write(
+            '\n'
+            '/* <opcbuiltins.h> */\n'
+            '/*   name,                builtin,               f1,                  f2,       f3,               i1,          i2,         o1,         o2,         m1, m2,         token */\n'
```
The table rows end with a "comment" field.   Should a "comment" label also be included in this file header?

WhoBrokeTheBuild replied:
The columns are for things inside the `OPC()` macro, and the comment is just whatever was left as a comment on each line

WhoBrokeTheBuild committed:
Deleting opcbuiltins.h.. again

mwinkel-dev reviewed (approved):
This refactoring looks good.   Some of the improvements . . . 
- opcodes.csv replaces the opcbuiltins.h and its OPC macros (which complicated debugging)
- clarity improved by renaming some files
- gen-messages-exceptions has a nice simplification for extracting the elements from the XML file, plus gets rid of the unused "msglistm" argument
- deleted the opcbuiltins.h file (that was added by the rebase)

Furthermore, prior versions of this PR successfully built and passed the automated tests.  And another core developer has already reviewed this.

For all the above reasons, this PR is low risk.   Thus approving.
